### DISCUSSION TITLE
Can JoltPhysics.js replace ammo.js at present ?

### DISCUSSION THREAD

**Discussion by finscn** [jrouwe/JoltPhysics.js#28] [General]
I want to replace ammo.js with JoltPhysics.js in my web game.
But I don't know   Whether current version of JoltPhysics.js  covers all the functions/features of  ammo.js (Bullet)

I think the native version JoltPhysics is good enough to replace Bullet.
What about JoltPhysics.js ?

**Comment by jrouwe** [comment]
I think in terms of features Jolt and Bullet are quite close. I did a quick scan through [ammo.idl](https://github.com/kripken/ammo.js/blob/main/ammo.idl) to see what the differences are between ammo.js and JoltPhysics.js in term of features. I think the main things that are missing are:

* ~~Collision tests (cast ray, cast shape, collide shape and broad phase collide or as they are called in Bullet: rayTest, convexSweepTest, contactTest and contactPairTest)~~ Done
* ~~Custom collision groups (right now only 2 groups are exposed in the JS interface)~~ Done
* ~~6DOF Constraint (all other constraints that ammo.js supports have been exposed)~~ Done
* ~~Vehicle support~~ Done

Exposing these is not a lot of work as it is mostly a matter of typing the class/function names in [JoltJS.idl](https://github.com/jrouwe/JoltPhysics.js/blob/main/JoltJS.idl).

**Comment by finscn** [comment]
Thanks. 
I hope  Jolt.js would expost these features in the feature .

**Comment by jrouwe** [comment]
I just released a version that fixes the first item.

**Comment by jrouwe** [comment]
6DOF constraint is exposed now too

**Comment by finscn** [comment]
Hi @jrouwe ,  Is there any plan for  `Custom collision groups` & `Vehicle support` ?

**Reply by jrouwe** [reply]
Yes to both, vehicle support should be simple but I haven't decided how to expose the custom collision groups yet.

**Reply by jrouwe** [reply]
Vehicle support has been implemented, I hope to implement custom collision groups soon

**Reply by jrouwe** [reply]
And the [latest release](https://github.com/jrouwe/JoltPhysics.js/releases/tag/0.10.0) now also supports defining your own layers.
